maas/client/viscera/boot_resources.py

```diff
@@ -0,0 +1,123 @@
+"""Objects for boot resources."""
+
+__all__ = [
+    "BootResource",
+    "BootResources",
+]
+
+from typing import Mapping
+
+from . import (
+    check,
+    check_optional,
+    Object,
+    ObjectField,
+    ObjectSet,
+    ObjectType,
+    parse_timestamp,
+)
+
+
+class BootResourceMappingType(ObjectType):
+    """Metaclass for `BootResourceSet` and `BootResourceFile`."""
+
+    def mapping(cls, data):
+        """Convert `data` mapping to mapping for class."""
+        if not isinstance(data, Mapping):
+            raise TypeError(
+                "data must be a mapping, not %s"
+                % type(data).__name__)
+        return {
+            key: cls(value)
+            for key, value in data.items()
+        }
```

Comment:
This could be generalised:

```python
def mapping_of(cls):
    """Expects a mapping from some key to data for `cls` instances."""
    def mapper(data):
        if not isinstance(data, Mapping):
            raise TypeError(
                "data must be a mapping, not %s"
                % type(data).__name__)
        return {
            key: cls(value)
            for key, value in data.items()
        }
    return mapper
```

and used like:

```python
    files = ObjectField.Checked(
        "files", mapping_of(BootResourceFile), default={})
```

Perhaps make this field `readonly=True`.

Note that the mutable value for `default` is dangerous.